### PR TITLE
TP Adjustment

### PR DESCRIPTION
**Pull request by aarongchan**
Notes:
Not too sure about the learning rate applied to the gradient, I chose 10 because I saw the gradients could be tiny and didn't want the updates to take a while, but not sure if you have a preferred learning rate.

Steps to run:
python3 nn/tp_adjustment.py -gd $DATA_DIR/graph -cd $DATA_DIR/ckpts/split#/seed -tp designs/ibex_v1/data/generated/tests/#.yaml -td $DATA_DIR/tp_cov -cp #

for me the run command was:
python3 nn/tp_adjustment.py -gd toAaron/graph/ -cd data/ckpts/split_0.7,0.15,.15/seed-123/ -tp designs/ibex_v1/data/generated/tests/00000.yaml -td toAaron/tp_cov -cp 7

### CODE REVIEW

File: nn/tp_adjustment.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  model_config = json.load(model_config_json)
+  model_config = model_config["model_config"]
+  model = get_d2v_model(**model_config)
+  model.load_weights(os.path.join(ckpt_dir, "model.ckpt"))
+
+  coverage = 0
+  prev_coverage = 1
+  while abs(prev_coverage - coverage) > .01:
+    prev_coverage = coverage
+    x = update_tp(tp, 1, graphs, bvocab)
```
Comment (insoochung) [inline nn/tp_adjustment.py]: 1 should be cp_idx?

File: nn/tp_adjustment.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  """
+  gen_opts_str = (f"test: {adjusted_test_name} \n") + gen_opts_str
+  with open(f"{adjusted_test_name}.yaml", "a") as f:
+    f.write(gen_opts_str)
+
+def adjust_tp(graph_dir, ckpt_dir, test_parameters, tp_cov_dir, cp_idx):
+  vocab_files = glob(os.path.join(tp_cov_dir, "vocab.*.yaml"))
+  vocab = TestParameterVocab(vocab_filepath=vocab_files[1])
+  tp_path = test_parameters
+  tp = vocab.vectorize_test(tp_path)
```
Comment (insoochung) [inline nn/tp_adjustment.py]: I think we should start from a random vector instead of a real YAML.

Comment (insoochung) [reply nn/tp_adjustment.py]: So the flow will focus on which coverpoint to hit and not worry too much about the "starting point".

File: nn/tp_adjustment.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-gd", "--graph_dir", type=str, required=True,
+                      help="Directory of the graph dataset.")
+    parser.add_argument("-cd", "--ckpt_dir", type=str, required=True,
+                      help="Directory to save the checkpoints.")
+    parser.add_argument("-tp", "--test_parameters", type=str, required=True,
+                      help="Test paramters .yaml path")         
+    parser.add_argument("-td", "--tp_cov_dir", type=str,
+                      help="Directory of the test parameter coverage dataset.")
+    parser.add_argument("-cp", "--cp_idx", type=str,
```
Comment (insoochung) [inline nn/tp_adjustment.py]: type should be int

File: nn/tp_adjustment.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    mask = bvocab.get_mask(cp_idx, midx, mask_length=graphs[0].n_nodes)
+    mask = np.ma.make_mask(mask)
+    is_hit = np.array([1])
+    gidx = np.array([midx])
+    cp = np.array([cp_idx])
+
+    ret = {
+      "test_parameters": np.array([tp]).reshape(1, 58),
+      "is_hits": is_hit.reshape(1, 1),
+      "coverpoint_mask": mask.reshape(1, 740),
```
Comment (insoochung) [inline nn/tp_adjustment.py]: Could we get shape info from the vocab class instead of hard-coded numbers? This will break later on.

File: nn/tp_adjustment.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  coverage = 0
+  prev_coverage = 1
+  while abs(prev_coverage - coverage) > .01:
+    prev_coverage = coverage
+    x = update_tp(tp, 1, graphs, bvocab)
+    with tf.GradientTape() as g:
+      g.watch(x)
+      y = model.call(x)
+    coverage = y[[0]]
+    dy_dx = g.gradient(y, x)["test_parameters"]
```
Comment (insoochung) [inline nn/tp_adjustment.py]: Shouldn't this be:
```python
loss = 1 - y
dy_dx = g.gradient(loss, x)
```

**Comment by insoochung** [comment]
@AaronGChan, hey Aaron nice work overall. Please check some of the comments when you have time and let me know if you have any doubts.

File: nn/tp_adjustment.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    tp += dy_dx * 10
```
Comment (insoochung) [inline nn/tp_adjustment.py]: We should add a line to 1) confirm all tp are in range 0~1 and 2) bolean elements are rounded to 0 or 1.

File: nn/tp_adjustment.py [hunk 7]
```diff
@@ -50,32 +51,45 @@ def adjust_tp(graph_dir, ckpt_dir, test_parameters, tp_cov_dir, cp_idx):
...
-      g.watch(x)
-      y = model.call(x)
-    coverage = y[[0]]
-    dy_dx = g.gradient(y, x)["test_parameters"]
-    tp += dy_dx * 10
-
-  tp = tp.numpy()[0]
+  is_hit = 0
+  lr = 1
+  step_threshold = 50
```
Comment (insoochung) [inline nn/tp_adjustment.py]: One last thing, please have these passable by flags and their defaults be 1 and 50, thx.

File: data/utils.py [hunk 8]
```diff
@@ -325,6 +325,27 @@ def generate_from_test_template(self):
...
+      idx = token["idx"]
+      if "name" in key:
+        begin, end = key.split("+")
+        test_parameters_vec[begin] = end
+        continue 
+      if token["is_one_hot"]:
+        assert token["type"] == "choice"
+        key, one_hot_val = key.split("+")
+        test_parameters_vec[key] = int(tp[idx] == one_hot_val)
+        continue
```
Comment (insoochung) [inline data/utils.py]: This doesn't prevent cases

test_parameters_vec["seq_type+type1"] == 0
test_parameters_vec["seq_type+type2"] == 1
test_parameters_vec["seq_type+type3"] == 1

Only one of these positions should be 1

File: data/utils.py [hunk 9]
```diff
@@ -325,6 +325,27 @@ def generate_from_test_template(self):
     self.tokens = vocab
     self.meta = test_parameters
 
+
+  def de_normalize(self, tp):
```
Comment (insoochung) [inline data/utils.py]: This function does not handle boolean cases.

File: nn/tp_adjustment.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    else:
+      x = update_tp(tp, cp_idx, graphs, bvocab, tp_size)
+      with tf.GradientTape() as g:
+        g.watch(x)
+        y = model.call(x)
+        loss = y[[0]] - 1
+      is_hit = y[[0]]
+      dy_dx = g.gradient(loss, x)["test_parameters"]
+      tp += dy_dx * lr
+  tp = [1 if i > 1 else 0 if i < 0 else i for i in tp.numpy()[0]]
```
Comment (insoochung) [inline nn/tp_adjustment.py]: This works for int but for other types, this is too loose. TP vocab should handle this and enforce better constraints.

**Comment by insoochung** [comment]
@AaronGChan, I've fixed some of the issues I've commented on. I'll merge this, please continue from the `main` branch.
